### PR TITLE
[DirectX][SPIRV] Allow configuration of instanced draw parameters

### LINKED ISSUE #1444: [DirectX][SPRIV] Allow configuration of instanced draw parameters
Required for testing `SV_InstanceID`, `SV_StartVertexLocation`, and `SV_StartInstanceLocation`.

Current status: the test harness has all of `InstanceCount`, `FirstVertex`, and `FirstInstance` hardcoded at the
call sites that invoke `Encoder::drawInstanced`. We will need to add yaml config to pass these parameters through so they can be specified.

AC:

- [ ] Add YAML config for `InstanceCount`, `FirstVertex` and `FirstInstance` to `DispatchParametersSet`
- [ ] Pass all three values through at the three `drawInstanced` call sites
- [ ] Add an `SV_InstanceID` test to check this is assigned correctly using a readback buffer
- [ ] Add `SV_StartVertexLocation` / `SV_StartInstanceLocation` test to check these values assigned correctly using a readback buffer

### PR DESCRIPTION
**Pull request by jzakharovnv**
Fixes #1444 and [hlsl wg issue](https://github.com/llvm/wg-hlsl/issues/152)

Testing for [SV_InstanceID](https://github.com/llvm/llvm-project/issues/212572)

Marked xfail until [clang SV_InstanceID implementation](https://github.com/llvm/llvm-project/pull/221091) and [clang SV refactor](https://github.com/llvm/llvm-project/pull/218058) land

@dnovillo @pow2clk

Assisted-by: LLM

### CODE REVIEW

File: lib/Support/Pipeline.cpp [hunk 1]
```diff
@@ -962,6 +965,14 @@ llvm::Error offloadtest::Pipeline::validateDispatchParameters() {
       return llvm::createStringError(
           "DispatchParameters.VertexCount set on a Compute or Mesh Shader "
           "pipeline. Only allowed on a TraditionalRaster pipeline.");
+    if (DispatchParameters.InstanceCount != 1 ||
+        DispatchParameters.FirstVertex != 0 ||
+        DispatchParameters.FirstInstance != 0)
```
Comment (dnovillo) [inline lib/Support/Pipeline.cpp:970]: RT dispatch does not consume these fields. Could you reject nondefault `InstanceCount`, `FirstVertex`, and `FirstInstance` values in the `RayTracing` case and add a parsing test that verifies the error?

File: test/Feature/Semantics/InstanceID.test [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#--- pipeline.yaml
+---
+Shaders:
+  - Stage: Vertex
+    Entry: main
+  - Stage: Pixel
+    Entry: main
+DispatchParameters:
+  VertexCount: 3
+  InstanceCount: 4
```
Comment (dnovillo) [inline test/Feature/Semantics/InstanceID.test:46]: `FirstVertex` and `FirstInstance` do not seem covered. Could you add a test with non-zero `FirstVertex` and `FirstInstance`? When reading back start vertex location and start instance location, `SV_InstanceID` should remain zero-based.

**Review by dnovillo — CHANGES_REQUESTED** [review]
Thanks for these tests. I'm not very experienced with OTS, so I may have missed a few things in my review.

File: test/Feature/Semantics/InstanceID.test (hunk 2 above)
Comment (jzakharovnv) [reply test/Feature/Semantics/InstanceID.test:46]: Yeah, let me add testing for those in the following days.

File: lib/Support/Pipeline.cpp (hunk 1 above)
Comment (jzakharovnv) [reply lib/Support/Pipeline.cpp:970]: Yes, let me add this too with the next commit

File: test/Feature/Semantics/InstanceID.test (hunk 2 above)
Comment (jzakharovnv) [reply test/Feature/Semantics/InstanceID.test:51]: Fixed

File: lib/Support/Pipeline.cpp (hunk 1 above)
Comment (jzakharovnv) [reply lib/Support/Pipeline.cpp:970]: Fixed

**Review by dnovillo — APPROVED** [review]
LGTM. Thanks.

**Comment by jzakharovnv** [comment]
@bogner Could you take a look? Thanks!

**Review by spall — APPROVED** [review]
lgtm

File: lib/Support/Pipeline.cpp [hunk 3]
```diff
@@ -962,6 +965,14 @@ llvm::Error offloadtest::Pipeline::validateDispatchParameters() {
...
+      return llvm::createStringError(
+          "DispatchParameters.InstanceCount/FirstVertex/FirstInstance set on a "
+          "Compute or Mesh Shader pipeline. Only allowed on a "
+          "TraditionalRaster "
+          "pipeline.");
```
Comment (inbelic) [inline lib/Support/Pipeline.cpp:975]: nit: weird string formatting

File: test/Feature/Semantics/InstanceID.test [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  - Result: InstanceIDsOffset
+    Rule: BufferExact
+    Actual: Output
+    Expected: Output_Expected
+...
+#--- end
+
+# SV_StartVertexLocation and SV_StartInstanceLocation not yet
+# implemented in the Clang-DXC SPIRV or DXIL backend.
+# XFAIL: Clang
```
Comment (inbelic) [inline test/Feature/Semantics/InstanceID.test:169]: nit: Ideally this would look like:
```suggestion
# Unimplemented: ISSUE LINK and ISSUE LINK
# XFAIL: Clang
```

File: test/Feature/Semantics/InstanceID.test
Comment (inbelic) [inline test/Feature/Semantics/InstanceID.test:1]: Can we also add tests that use the other semantics listed:  `SV_StartVertexLocation` and `SV_StartInstanceLocation`. Ideally in new test files. It might also be a little more clear to split this test file into the 2 pipeline cases

**Review by inbelic — COMMENTED** [review]
Mostly LGTM with some small requests for changes

Small note wrt the pr description, this will not pass on clang until the issues mentioned and https://github.com/llvm/llvm-project/issues/90504 lands. I'll be working on that this week

File: lib/Support/Pipeline.cpp (hunk 3 above)
Comment (jzakharovnv) [reply lib/Support/Pipeline.cpp:975]: Lemme fix

File: test/Feature/Semantics/InstanceID.test (hunk 4 above)
Comment (jzakharovnv) [reply test/Feature/Semantics/InstanceID.test:169]: Good suggestion, will add

File: test/Feature/Semantics/InstanceID.test
Comment (jzakharovnv) [reply test/Feature/Semantics/InstanceID.test:1]: Yeah, I can do that with the next commit

**Comment by jzakharovnv** [comment]
@inbelic Implemented your feedback, it also looks like it is passing clang- at least locally.
